File: LinearList/code/php/SequenceList.php
<?php
/**
 * 顺序表的基本操作实现
 * @author: phachon@163.com
 */

class SequenceList {

    /**
     * 默认最大容量
     */
    const Max = 100;

    /**
     * 顺序表的最大容量
     * @var int
     */
    private $_max = 0;

    /**
     * 顺序表元素个数
     * @var int
     */
    private $_num = 0;


    /**
     * 数据元素
     * @var array
     */
    private $_data = NULL;

    /**
     * SequenceList constructor.
     * @param int $max
     */
    public function __construct($max = self::Max) {
        $this->_max = $max;
        $this->_num = 0;
        $this->_data = new SplFixedArray($max);
    }

    /**
     * 创建顺序表
     * @param $max
     * @return SequenceList
     */
    public static function create($max) {
        return new SequenceList($max);
    }

    /**
     * 添加数据元素到起始位置
     * @param string $val
     * @throws Exception
     */
    public function addFirst($val = '') {
        $this->add(0, $val);
    }

    /**
     * 添加数据元素到末尾位置
     * @param string $val
     * @throws Exception
     */
    public function addLast($val = '') {
        $this->add($this->_num, $val);
    }

    /**
     * 添加数据元素到任意位置
     * @param $index
     * @param string $val
     * @throws Exception
     */
    public function add($index, $val = '') {
        if (!is_integer($index)) {
            throw new Exception("Add failed, index must be a number!");
        }
        if ($index > ($this->_max - 1) || $index < 0) {
            throw new Exception("Add failed, index error!");
        }

        for ($i = $this->_num - 1; $i > $index; $i--) {
            $this->_data[$i] = $this->_data[$i-1];
        }

        $this->_data[$index] = $val;
        $this->_num += 1;
    }

    /**
     * 查找指定位置的数据元素
     * @param $index
     * @return mixed
     * @throws Exception
     */
    public function get($index) {
        if (!is_integer($index)) {
            throw new Exception("Get failed, index must be a number!");
        }
        if ($index > ($this->_num - 1) || $index < 0) {
            throw new Exception("Get failed, index error!");
        }

        return $this->_data[$index];
    }

    /**
     * 修改指定位置的数据元素
     * @param $index
     * @param string $val
     * @throws Exception
     */
    public function set($index, $val = '') {
        if (!is_integer($index)) {
            throw new Exception("Set failed, index must be a number!");
        }
        if ($index > ($this->_num - 1) || $index < 0) {
            throw new Exception("Set failed, index error!");
        }

        $this->_data[$index] = $val;
    }

    /**
     * 删除指定位置的元素
     * @param $index
     * @throws Exception
     */
    public function del($index) {
        if (!is_integer($index)) {
            throw new Exception("Del failed, index must be a number!");
        }
        if ($index > ($this->_num - 1) || $index < 0) {
            throw new Exception("Del failed, index error!");
        }

        for ($i = $index; $i < $this->_num; $i++) {
            $this->_data[$i] = $this->_data[$i+1];
        }

        $this->_num -= 1;
    }

    /**
     * 顺序表是否为空
     * @return bool
     */
    public function isEmpty() {
        return empty($this->_num);
    }

    /**
     * 顺序表是否已满
     */
    public function isFull() {
        return $this->_num == $this->_max;
    }

    /**
     * 遍历输出数据元素
     */
    public function printlnItem() {
        $printStr = "[";
        foreach ($this->_data as $item) {
            if ($item == "") {
                continue;
            }
            $printStr = $printStr.$item.",";
        }
        $printStr = rtrim($printStr, ",");
        $printStr .= "]";
        echo $printStr."\r\n";
    }

    /**
     * 获取顺序表的元素个数
     * @return int
     */
    public function count() {
        return $this->_num;
    }

    /**
     * 元素是否存在
     * @param string $val
     * @return bool
     */
    public function isExist($val = '') {
        foreach ($this->_data as $item) {
            if ($item == $val) {
                return true;
            }
        }
        return false;
    }

    /**
     * 根据值查找元素的位置
     * @param string $val
     * @return int
     */
    public function find($val = '') {
        foreach ($this->_data as $index => $item) {
            if ($item == $val) {
                return $index;
            }
        }
        return -1;
    }

    /**
     * 销毁顺序表
     */
    public function destroy() {
        $this->__construct();
    }
}

File: LinearList/code/golang/sequence_list.go
package golang

import (
	"errors"
	"fmt"
	"strings"
)

// 顺序表的操作实现

type SequenceList struct {
	// 最大容量
	max int

	// 长度
	length int

	// 数据
	data []interface{}
}

// 创建顺序表
func NewSequenceList(max int) *SequenceList {
	return &SequenceList{
		max: max,
		length: 0,
		data: make([]interface{}, max),
	}
}

// 添加元素到起始位置
func (s *SequenceList) AddFirst(val interface{}) error {
	return s.add(0, val)
}

// 添加元素到末尾位置
func (s *SequenceList) AddLast(val interface{}) error {
	return s.add(s.length - 1, val)
}

// 添加元素到指定位置 index
func (s *SequenceList) add(index int, val interface{}) error {
	if s.length == s.max {
		return errors.New("Add failed, list is full! ")
	}
	if index < 0 || index > s.max {
		return errors.New("Add failed, index out of range! ")
	}
	for i := s.length; i >= index; i-- {
		s.data[i+1] = s.data[i]
	}
	s.data[index] = val
	s.length++

	return nil
}

// 获取指定位置的数据元素
func (s *SequenceList) Get(index int) (interface{}, error) {
	if index < 0 || index > s.length {
		return nil, errors.New("Get failed, index out of range! ")
	}
	return s.data[index], nil
}

// 修改指定位置的元素
func (s *SequenceList) Set(index int, val interface{}) error {
	if index < 0 || index > s.length {
		return errors.New("Set failed, index out of range! ")
	}
	s.data[index] = val
	return nil
}

// 删除指定位置的元素
func (s *SequenceList) Del(index int) error {
	if index < 0 || index > s.length {
		return errors.New("Del failed, index out of range! ")
	}

	for i := index; i < s.length; i++ {
		s.data[i] = s.data[i+1]
	}
	s.data[s.length - 1] = nil
	s.length --
}

// 顺序表是否为空
func (s *SequenceList) IsEmpty() bool {
	return s.length == 0
}

// 顺序表是否已满
func (s *SequenceList) IsFull() bool {
	return s.length == s.max
}

// 遍历输出数据元素
func (s *SequenceList) PrintlnItem() {
	str := ""
	for _, val := range s.data {
		str += fmt.Sprintf("%v,", val)
	}
	str = strings.TrimRight(str, ",")
	fmt.Println(str)
}

// 获取顺序表的元素个数
func (s *SequenceList) Count() int {
	return s.length
}

// 元素是否存在
func (s *SequenceList) IsExist(val interface{}) bool {

	for i := 0; i < s.length; i++ {
		if s.data[i] == val {
			return true
		}
	}
	return false
}

// 根据值查找元素的位置
func (s *SequenceList) Find(val interface{}) int {
	for i := 0; i < s.length; i++ {
		if s.data[i] == val {
			return i
		}
	}
	return -1
}

// 销毁顺序表
func (s *SequenceList) Destroy() {
	for i := 0; i < s.length; i++ {
		s.data[i] = nil
	}
	s.max = 0
	s.length = 0
}

// 清空顺序表
func (s *SequenceList) Clear() {
	for i := 0; i < s.length; i++ {
		s.data[i] = nil
	}
	s.length = 0
}